cpp/test/distance/sparse_l2_nn.cu

```diff
@@ -0,0 +1,494 @@
+              << p.num_groups
+              << ", "
+                 "seed: "
+              << p.seed
+              << ", "
+                 "tol: "
+              << p.tolerance;
+  }
+};
+
+__global__ void init_adj(
+  int m, int n, int num_groups, AdjacencyPattern pattern, bool* adj, int* group_idxs)
+{
+  for (int i = blockIdx.y * blockDim.y + threadIdx.y; i < num_groups; i += blockDim.y * gridDim.y) {
+    for (int j = blockIdx.x * blockDim.x + threadIdx.x; j < m; j += blockDim.x * gridDim.x) {
+      switch (pattern) {
+        case checkerboard: adj[i * m + j] = (i + j) % 2; break;
+        case checkerboard_4: adj[i * m + j] = (i + (j / 4)) % 2; break;
+        case checkerboard_64: adj[i * m + j] = (i + (j / 64)) % 2; break;
+        case all_true: adj[i * m + j] = true; break;
+        case all_false: adj[i * m + j] = false; break;
+        default: assert(false && "unknown pattern");
+      }
+    }
+  }
+  // Each group is of size n / num_groups.
+  //
+  // - group_idxs[j] indicates the start of group j + 1 (i.e. is the inclusive
+  // scan of the group lengths)
+  //
+  // - The first group always starts at index zero, so we do not store it.
+  //
+  // - The group_idxs[num_groups - 1] should always equal n.
+
+  if (blockIdx.y == 0 && threadIdx.y == 0) {
+    const int j_stride = blockDim.x * gridDim.x;
+    for (int j = blockIdx.x * blockDim.x + threadIdx.x; j < num_groups; j += j_stride) {
+      group_idxs[j] = (j + 1) * (n / num_groups);
+    }
+    group_idxs[num_groups - 1] = n;
+  }
+}
+
+template <typename DataT, bool Sqrt>
+class SparseL2NNTest : public ::testing::TestWithParam<Inputs<DataT>> {
+ public:
+  SparseL2NNTest()
+    : params(::testing::TestWithParam<Inputs<DataT>>::GetParam()),
+      stream(handle.get_stream()),
+      x(params.m * params.k, stream),
+      y(params.n * params.k, stream),
+      adj(params.m * params.num_groups, stream),
+      group_idxs(params.num_groups, stream),
+      xn(params.m, stream),
+      yn(params.n, stream),
+      min(params.m, stream),
+      min_ref(params.m, stream),
+      workspace(params.m * sizeof(int), stream)
+  {
+  }
+
+ protected:
+  void SetUp() override
+  {
+    raft::random::RngState r(params.seed);
+    int m          = params.m;
+    int n          = params.n;
+    int k          = params.k;
+    int num_groups = params.num_groups;
+    uniform(handle, r, x.data(), m * k, DataT(-1.0), DataT(1.0));
+    uniform(handle, r, y.data(), n * k, DataT(-1.0), DataT(1.0));
+
+    dim3 block(32, 32);
+    dim3 grid(10, 10);
+    init_adj<<<grid, block, 0, stream>>>(
+      m, n, num_groups, params.pattern, adj.data(), group_idxs.data());
+    RAFT_CUDA_TRY(cudaGetLastError());
+
+    generateGoldenResult();
+    raft::linalg::rowNorm(xn.data(), x.data(), k, m, raft::linalg::L2Norm, true, stream);
+    raft::linalg::rowNorm(yn.data(), y.data(), k, n, raft::linalg::L2Norm, true, stream);
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
```

Comment:
Nitpick: `handle.sync_stream()` is a little safer and less error-prone, though I know we set a stream explicitly in most of the other googletests.